### PR TITLE
cli: Propagating errors from fallible `CliToSdk` conversions

### LINKED ISSUE #1815: Converting from CLI to SDK types should be fallible
Some conversions from CLI to SDK types require file I/O. Of course, reading from and writing to files can fail. However, currently `CliToSdk` does not have a way to forward errors to a higher stack frame. This encourages panicking in the CLI, which is bad UX.

The suggested approach is to turn `CliToSdk` into something like:

```rs
/// Convert CLI args to SDK args, with contextual data.
pub trait CliToSdk<SDK>: Args {
    /// Error type for fallible operations.
    type Error;

    /// Convert CLI args to SDK args, with contextual data.
    fn to_sdk(self, ctx: &mut Context) -> Result<SDK, Self::Error>;

    /// Convert CLI args to SDK args, with contextual data.
    fn infallible_to_sdk(self, ctx: &mut Context) -> SDK
    where
        // NB: matching on the type with `==` is not stable AFAIK,
        // so we use `Into<T>`, which is implemented for any `T`
        Self::Error: Into<std::convert::Infallible>,
    {
        match self.to_sdk(ctx).map_err(Into::into) {
            Ok(x) => x,
            Err(e) => match e {},
        }
    }
}
```

Whenever errors are not expected, it is possible to return a `Result<T, std::convert::Infallible>`, which has 0 cost from a runtime perspective.

### PR DESCRIPTION
**Pull request by phasewalk1**
## Describe your changes

This PR closes #1815. I've refactored the `CliToSdk` trait to propagate errors (such as I/O) from _fallible_ conversions between SDK and CLI types, this way we can avoid panics.

## Indicate on which release or other PRs this topic is based on
Based on `v0.31.8`

## Checklist before merging to `draft`
- [x] I have added a changelog
- [x] Git history is in acceptable state

### CODE REVIEW

File: crates/apps/src/lib/cli.rs [hunk 1]
```diff
@@ -6090,7 +6390,19 @@ pub mod args {
...
+                            .map_err(|e| {
+                                std::io::Error::new(
+                                    std::io::ErrorKind::InvalidInput,
+                                    format!(
+                                        "Error reading signature file: {}",
+                                        e
+                                    ),
+                                )
+                            })
+                            .unwrap()
```
Comment (sug0) [inline crates/apps/src/lib/cli.rs:6404]:
```suggestion
```

File: crates/apps/src/lib/cli.rs [hunk 2]
```diff
@@ -6090,7 +6390,19 @@ pub mod args {
...
+                    })
                     .collect(),
```
Comment (sug0) [inline crates/apps/src/lib/cli.rs:6406]:
```suggestion
                    .collect()?,
```
you can collect a `Result<C, _>` for some collection `C` from an iterator yielding `Result` values

File: crates/apps/src/lib/cli.rs [hunk 3]
```diff
@@ -4618,16 +4705,21 @@ pub mod args {
...
+
         fn to_sdk(
             self,
             ctx: &mut Context,
-        ) -> UpdateStewardCommission<SdkTypes> {
-            UpdateStewardCommission::<SdkTypes> {
-                tx: self.tx.to_sdk(ctx),
+        ) -> Result<UpdateStewardCommission<SdkTypes>, Self::Error> {
+            let tx = self.tx.to_sdk(ctx)?;
+            let _commission = std::fs::read(self.commission)?;
```
Comment (sug0) [inline crates/apps/src/lib/cli.rs:4715]:
```suggestion
            let commission = std::fs::read(self.commission)?;
```
nit: why did you use an underscore for the identifier here?

File: crates/apps/src/lib/cli.rs [hunk 4]
```diff
@@ -4744,15 +4849,23 @@ pub mod args {
...
-                tx: self.tx.to_sdk(ctx),
-                proposal_data: std::fs::read(self.proposal_data).expect(""),
+        type Error = std::io::Error;
+
+        fn to_sdk(
+            self,
+            ctx: &mut Context,
+        ) -> Result<InitProposal<SdkTypes>, Self::Error> {
+            let tx = self.tx.to_sdk(ctx)?;
+            let _proposal_data = std::fs::read(self.proposal_data)?;
```
Comment (sug0) [inline crates/apps/src/lib/cli.rs:4859]:
```suggestion
            let proposal_data = std::fs::read(self.proposal_data)?;
```

File: crates/apps/src/lib/cli.rs [hunk 5]
```diff
@@ -4830,19 +4943,26 @@ pub mod args {
...
+        type Error = std::io::Error;
+
+        fn to_sdk(
+            self,
+            ctx: &mut Context,
+        ) -> Result<VoteProposal<SdkTypes>, Self::Error> {
+            let tx = self.tx.to_sdk(ctx)?;
+            let _proposal_data = self.proposal_data.map(|path| {
+                std::fs::read(path).expect("Should be able to read the file.")
+            });
```
Comment (sug0) [inline crates/apps/src/lib/cli.rs:4955]:
```suggestion
            let proposal_data = self
                .proposal_data
                .map(std::fs::read)
                .transpose()?;
```

**Review by sug0 — CHANGES_REQUESTED** [review]
hey. this is a great effort! thanks for submitting this patch!

I just have a couple of suggestions:

**Comment by phasewalk1** [comment]
> hey. this is a great effort! thanks for submitting this patch!
> 
> I just have a couple of suggestions:

hey thank you for the feedback. should I commit each suggested change individually?

**Comment by sug0** [comment]
I generally prefer `fixup!` commits then a rebase, but a single commit with all the review changes is fine too

File: crates/apps/src/lib/cli.rs (hunk 2 above)
Comment (sug0) [reply crates/apps/src/lib/cli.rs:6406]:
```rs
.collect::<Result<Vec<_>, _>>
```

**Comment by phasewalk1** [comment]
I noticed that my fork is now quite behind the main branch (48 commits behind), and some CI tests are failing after I addressed your suggested changes. From my understanding, it's recommended to sync my fork with the main branch and rebase my changes on top of the latest commits.

However, I haven't had to do this process before (I'm an open source noob), and I'm a bit unsure about the proper steps. I want to ensure that I don't introduce any unintended changes or mess up the commit history. Could you please provide me with some guidance on how to proceed? Thank you 🙏 @sug0

**Comment by sug0** [comment]
hey @phasewalk1. yes, CI is probably broken on your current history, since we made changes in `main`. it should be pretty straight forward to do a rebase on `base`. run the following commands:
```
git remote add upstream https://github.com/anoma/namada
git fetch upstream
git rebase upstream/base
```
if you run into conflicts, I recommend installing Sublime Merge on macOS/Windows/Linux, or Meld on Windows/Linux to resolve them.

cheers,
-- sugo

**Comment by phasewalk1** [comment]
@sug0 cool, no conflicts. do i just force push to my remote now?

**Comment by sug0** [comment]
yup

**Review by tzemanovic — APPROVED** [review]

**Comment by phasewalk1** [comment]
hey @tzemanovic! I saw these changes got approved, should I rebase again and update my branch?

**Comment by tzemanovic** [comment]
> hey @tzemanovic! I saw these changes got approved, should I rebase again and update my branch?

pls do just to get a CI run

**Comment by phasewalk1** [comment]
@brentstone I'm running into some merge conflicts that are giving me a bit of trouble. I'm a little pressed for time right now, as I have a couple math exams coming up that are taking priority. I'll have more time on Friday to fix this, but in the meantime, I can give you write access to my fork if you want to ship it sooner rather than later.

**Comment by phasewalk1** [comment]
Following up on this- the conflicts are a bit pesky; I want to make sure I'm accepting the right changes. Happy to collab on this or provide write access to the fork, but as it stands I could use a little guidance. Thanks for your patience.

**Pull request merged**